Lib/fractions.py

```diff
@@ -567,11 +567,15 @@ def __hash__(self):
         # dinv is the inverse of self._denominator modulo the prime
         # _PyHASH_MODULUS, or 0 if self._denominator is divisible by
         # _PyHASH_MODULUS.
-        dinv = pow(self._denominator, _PyHASH_MODULUS - 2, _PyHASH_MODULUS)
+        try:
+            dinv =  pow(self._denominator, -1, _PyHASH_MODULUS)
+        except ValueError:
+            # ValueError means there is no modular inverse
+            dinv = 0
         if not dinv:
             hash_ = _PyHASH_INF
         else:
-            hash_ = abs(self._numerator) * dinv % _PyHASH_MODULUS
+            hash_ = hash(abs(self._numerator)) * dinv % _PyHASH_MODULUS
         result = hash_ if self >= 0 else -hash_
```

Comment:
Instead of `self >= 0` you can use `self._numerator >= 0`.

Reply:
Done.